File: benchmarks/fixtures/webapp_go/internal/routes/auth_routes.go
package routes

import (
    "webapp_go/internal/auth"
    "webapp_go/internal/services"
    "webapp_go/internal/database"
    "webapp_go/pkg/logger"
)

var authRouteLog = logger.GetLogger("routes.auth")

// LoginHandler handles login requests.
func LoginHandler(request map[string]interface{}) (map[string]interface{}, error) {
    authRouteLog.Info("Login request received")

    email, _ := request["email"].(string)
    password, _ := request["password"].(string)

    db := database.NewDatabaseConnection("localhost", 5432, "app", "user")
    authSvc := services.NewAuthenticationService(db)

    token, err := authSvc.Authenticate(email, password)
    if err != nil {
        authRouteLog.Error("Login failed: %v", err)
        return map[string]interface{}{"error": err.Error()}, err
    }

    authRouteLog.Info("Login successful for: %s", email)
    return map[string]interface{}{
        "token": token,
        "user":  email,
    }, nil
}

// LogoutHandler handles logout requests.
func LogoutHandler(request map[string]interface{}) (map[string]interface{}, error) {
    authRouteLog.Info("Logout request received")
    headers, _ := request["headers"].(map[string]string)
    token := auth.ExtractToken(headers)
    err := auth.RevokeToken(token)
    if err != nil {
        authRouteLog.Error("Logout failed: %v", err)
        return nil, err
    }
    authRouteLog.Info("Logout successful")
    return map[string]interface{}{"status": "logged_out"}, nil
}

// RefreshHandler handles token refresh requests.
func RefreshHandler(request map[string]interface{}) (map[string]interface{}, error) {
    authRouteLog.Info("Token refresh request")
    headers, _ := request["headers"].(map[string]string)
    oldToken := auth.ExtractToken(headers)
    newToken, err := auth.RefreshToken(oldToken)
    if err != nil {
        authRouteLog.Error("Token refresh failed: %v", err)
        return nil, err
    }
    return map[string]interface{}{"token": newToken}, nil
}

File: benchmarks/fixtures/webapp_c/routes/auth_routes.c
#include <stddef.h>

#include "routes/auth_routes.h"

#include "database/database_connection.h"
#include "services/authentication_service.h"
#include "util/logger.h"

struct AuthRoutes auth_routes_new(const char *prefix)
{
    struct AuthRoutes routes;
    routes.prefix = prefix;
    return routes;
}

/// POST /login. Hop between the controller and the service layer.
const char *login_handler(struct AuthRoutes *routes, const char *email, const char *password)
{
    struct Logger log = get_logger("routes.auth");
    logger_info(&log, "Handling login request");
    (void)routes;

    struct LoginHandler request;
    request.email = email;
    request.password = password;
    request.issued_token = NULL;
    request.status = APP_OK;

    struct DatabaseConnection db = database_connection_new("localhost", 5432, "app");
    struct AuthenticationService service = authentication_service_new(&db);
    base_service_initialize(&service.base);

    request.issued_token = authenticate(&service, request.email, request.password);
    if (request.issued_token == NULL) {
        request.status = APP_ERROR_AUTHENTICATION;
        logger_warn(&log, app_error_name(request.status));
    }
    return request.issued_token;
}

enum AppError logout_handler(struct AuthRoutes *routes, const char *token)
{
    struct Logger log = get_logger("routes.auth");
    logger_info(&log, "Handling logout request");
    (void)routes;

    struct DatabaseConnection db = database_connection_new("localhost", 5432, "app");
    struct AuthenticationService service = authentication_service_new(&db);
    base_service_initialize(&service.base);

    return authentication_service_logout(&service, token);
}
